File: targeted/try-removing-query.py
import urlparse
import urllib
import itertools
from org.parosproxy.paros.model import Model
from org.parosproxy.paros.extension.history import ExtensionHistory
from org.parosproxy.paros.network import HttpSender

def logger(note):
	print("[try-removing-query] " +  str(note))

def _fix_dict(data):
	for key in data:
		val = data[key]
		if len(val) == 1:
			data[key] = val[0]
	return data

def _findsubsets(items, size):
    return list(set(itertools.combinations(items, size)))

def invokeWith(msg):
	# org.apache.commons.httpclient.URI	
	url = msg.getRequestHeader().getURI().toString()
	qry = urlparse.urlsplit(url).query
	qs = urlparse.parse_qs(qry)
	data = dict(qs)
		
	if not data:
		logger("No query ")
		return 

	# https://www.zaproxy.org/2.5/javadocs/org/parosproxy/paros/network/HttpSender.html#MANUAL_REQUEST_INITIATOR
	sender = HttpSender(Model.getSingleton().getOptionsParam().getConnectionParam(), True, 1)
	history = ExtensionHistory()
	all_params = data.keys()
	data = _fix_dict(data)
	query_sets = [[]] # Need the first query to be empty
	for i in range(1, len(all_params)):
    	query_sets = query_sets + _findsubsets(all_params, i)
	query_sets = [list(s) for s in query_sets]
    
    logger("Starting data")
	logger(data)
    logger("Trying " + str(len(query_sets)) + " different query combos ... ")

	for params in query_sets:
		query = {p: data[p] for p in params}
		logger("Querying with params: " + ", ".join(query))
		logger(query)
		clone = msg.cloneRequest()
		new_qs = urllib.urlencode(query)
		clone.getRequestHeader().getURI().setQuery(new_qs)
		sender.sendAndReceive(clone, True)
		logger("body: " + clone.getResponseBody().toString())
		history.addHistory(clone, 1)
		
	"""
	dialog = history.getResendDialog()
	dialog.setMessage(clone)
 	dialog.setVisible(True)
	"""

File: targeted/try-removing-query.js
var Model            = Java.type('org.parosproxy.paros.model.Model');
var ExtensionHistory = Java.type('org.parosproxy.paros.extension.history.ExtensionHistory');
var HttpSender       = Java.type('org.parosproxy.paros.network.HttpSender');

var sender  = new HttpSender(Model.getSingleton().getOptionsParam().getConnectionParam(), true, 1)
var history = new ExtensionHistory();

/*
// Keep for reference
var dialog = history.getResendDialog();
dialog.setMessage(clone);
dialog.setVisible(true)
*/

function logger() {
	print('[' + this['zap.script.name'] + '] ' + arguments[0]);
}

// https://stackoverflow.com/questions/1714786/query-string-encoding-of-a-javascript-object
function serialize(obj) {
	var str = [];
	for (var p in obj)
	  if (obj.hasOwnProperty(p)) {
		str.push(encodeURIComponent(p) + "=" + encodeURIComponent(obj[p]));
	  }
	return str.join("&");
  }

// https://stackoverflow.com/questions/2090551/parse-query-string-in-javascript
function parseQuery(queryString) {
    var query = {};
    var pairs = (queryString[0] === '?' ? queryString.substr(1) : queryString).split('&');
    for (var i = 0; i < pairs.length; i++) {
        var pair = pairs[i].split('=');
        query[decodeURIComponent(pair[0])] = decodeURIComponent(pair[1] || '');
    }
    return query;
}

// https://codereview.stackexchange.com/questions/7001/generating-all-combinations-of-an-array
function getCombinations(chars) {
	var result = [];
	var f = function(prefix, chars) {
		logger(prefix)
		for (var i = 0; i < chars.length; i++) {
			result.push(prefix.concat(chars[i]));
			f(prefix.concat(chars[i]), chars.slice(i + 1));
		}
	}
	f([], chars);
	return result;
}

function invokeWith(msg) {
	var query = msg.getRequestHeader().getURI().getQuery()
	if (query.length === 0) {
		logger("No query ")
		return;
	}

	var queryData = parseQuery(query);
	var allParams = Object.keys(queryData);
	var queryCombos = getCombinations(allParams);
	queryCombos.push([]);

	logger("Working through " + queryCombos.length + " combo(s)");
	for (var i in queryCombos) {
		var data = {};
		var params = queryCombos[i];

		for (var j in params) {
			var key = params[j];
			data[key] = queryData[key];
		}

		logger();
		logger("Sending params: " + params.join(", "));
		var newQuery = serialize(data);
		clone = msg.cloneRequest();
		clone.getRequestHeader().getURI().setQuery(newQuery);
		sender.sendAndReceive(clone, true);
		history.addHistory(clone, 1);
		logger("body: " + clone.getResponseBody().toString());
	}
}